### PR TITLE
Modified ModelB with Teacher Load

### PR DESCRIPTION
**Pull request by QiyouChang**
Main change: modify the model such that teacher model is now loaded in modelB instead of trained separately in modelB

.yaml: add load_model_path param
train.py: add config path 
base_class.py: add load function

### CODE REVIEW

**Review by baolef — CHANGES_REQUESTED** [review]
don't need to comment out train_teacher. git is considering it as changes

File: ModelB_KD/KD/base_class.py [hunk 1]
```diff
@@ -73,86 +73,96 @@ def __init__(
...
-            self.post_epoch_call(ep)
-
-        self.teacher_model.load_state_dict(self.best_teacher_model_weights)
-        if save_model:
-            torch.save(self.teacher_model.state_dict(), save_model_pth)
-        if plot_losses:
-            plt.plot(loss_arr)
+    def load_teacher(self, load_model_pth):
+        print("Loading Teacher")
+        self.teacher_model.load_state_dict(torch.load(load_model_pth))
```
Comment (baolef) [inline ModelB_KD/KD/base_class.py]: where do you get this from? have you verified it?

**Comment by QiyouChang** [comment]
Sure. Changes done!

File: configs/cifar10_vkd.yaml [hunk 2]
```diff
@@ -3,6 +3,7 @@ epochs: 100
 output_dir: save
 resume: null
 save_interval: 10
+load_model_path: models/teacher.pt
```
Comment (baolef) [inline configs/cifar10_vkd.yaml:6]: where does this come from?

**Review by baolef — COMMENTED** [review]
where do you get your loading code? is it verified?

File: configs/cifar10_vkd.yaml (hunk 2 above)
Comment (QiyouChang) [reply configs/cifar10_vkd.yaml:6]: This suppose to be the location of the teacher model. We can change it later

File: ModelB_KD/KD/base_class.py (hunk 1 above)
Comment (baolef) [reply ModelB_KD/KD/base_class.py]: have you verified it? If not, I believe it is wrong. Check my loading code for how to load a model

Comment (QiyouChang) [reply ModelB_KD/KD/base_class.py]: I wrote this. 
https://pytorch.org/tutorials/beginner/saving_loading_models.html

It is verified. I have run it locally seeing a large change in accuracy in first couple data -> model is loaded -> verified.

Comment (baolef) [reply ModelB_KD/KD/base_class.py]: ok, so you are not using my model weights, right?

Comment (baolef) [reply ModelB_KD/KD/base_class.py]: If you are using your own model weights, then where is the weight for the large teacher model? Has it been trained? Do you intend to retrain?

If you are using my weights, this won't load.

Comment (QiyouChang) [reply ModelB_KD/KD/base_class.py]: I am using the weights of the loaded teacher model. 

I am not sure why you say "it won't load" since you can see .weight parameter being initialized here as none zero values.  self.teacher_model is already initiated in the prepare function so why do you think it will not work?

Screenshot 2024-04-16 at 6 29 31 PM

Comment (baolef) [reply ModelB_KD/KD/base_class.py]: This is how to load modelA weights. If you use your code, you won't be able to use the weights of modelA.https://github.com/QiyouChang/TrainTimeKD/blob/main/ModelA_Vit/vit_utils/model.py#L15

If you use weights trained by yourself, then we need to make sure that the pretrained model is large enough and good enough.

**Review by baolef — COMMENTED** [review]
lgtm
